scripts/ci/pr-bot/processNewPrs.ts

```diff
@@ -0,0 +1,220 @@
+// 1) Checking if it needs processed
+// 2) Reminding reviewers if checks have completed (if they've subscribed to that)
+// 3) Picking/assigning reviewers
+// 4) Adding "Next Action: Reviewers label"
+// 5) Storing the state of the pull request/reviewers in a dedicated branch.
+async function processPull(
+  pull,
+  reviewerConfig: typeof ReviewerConfig,
+  stateClient: typeof PersistentState
+) {
+  let prState = await stateClient.getPrState(pull.number);
+  if (!needsProcessed(pull, prState)) {
+    return;
+  }
+
+  let checkState = await getChecksStatus(REPO_OWNER, REPO, pull.head.sha);
+
+  if (prState.remindAfterTestsPass && prState.remindAfterTestsPass.length > 0) {
+    return await remindIfChecksCompleted(
+      pull,
+      stateClient,
+      checkState,
+      prState
+    );
+  }
+
+  if (!checkState.succeeded) {
+    return await notifyChecksFailed(pull, stateClient, prState);
+  }
+  prState.commentedAboutFailingChecks = false;
+
+  // Pick reviewers to assign. Store them in reviewerStateToUpdate and update the prState object with those reviewers (and their associated labels)
+  let reviewerStateToUpdate = {};
+  const reviewersForLabels: { [key: string]: string[] } =
+    reviewerConfig.getReviewersForLabels(pull.labels, [pull.user.login]);
+  var labels = Object.keys(reviewersForLabels);
+  if (!labels || labels.length === 0) {
+    return;
+  }
+  for (let i = 0; i < labels.length; i++) {
+    let label = labels[i];
+    let availableReviewers = reviewersForLabels[label];
+    let reviewersState = await stateClient.getReviewersForLabelState(label);
+    let chosenReviewer = reviewersState.assignNextReviewer(availableReviewers);
+    reviewerStateToUpdate[label] = reviewersState;
+    prState.reviewersAssignedForLabels[label] = chosenReviewer;
+  }
+
+  console.log(`Assigning reviewers for PR ${pull.number}`);
+  await github.addPrComment(
+    pull.number,
+    commentStrings.assignReviewer(prState.reviewersAssignedForLabels)
+  );
+
+  github.nextActionReviewers(pull.number, pull.labels);
+  prState.nextAction = "Reviewers";
+
+  await stateClient.writePrState(pull.number, prState);
+  let labelsToUpdate = Object.keys(reviewerStateToUpdate);
+  for (let i = 0; i < labelsToUpdate.length; i++) {
+    let label = labelsToUpdate[i];
+    await stateClient.writeReviewersForLabelState(
+      label,
+      reviewerStateToUpdate[label]
+    );
+  }
+}
+
+async function processNewPrs() {
+  const githubClient = github.getGitHubClient();
+  let reviewerConfig = new ReviewerConfig(PATH_TO_CONFIG_FILE);
+  let stateClient = new PersistentState();
+
+  let openPulls = await githubClient.paginate(
+    "GET /repos/{owner}/{repo}/pulls",
+    {
+      owner: REPO_OWNER,
+      repo: REPO,
+    }
+  );
+
+  for (let i = 0; i < openPulls.length; i++) {
+    let pull = openPulls[i];
+    await processPull(pull, reviewerConfig, stateClient);
+  }
+}
+
+processNewPrs();
+
+export {};
```

Comment:
I've never seen this self executing pattern before. I think exporting `processNewPrs` and then calling it in the index.ts file would be more explicit. Hiding execution under an import could make debugging more difficult later.

Reply:
This is the standard pattern used for all first party GitHub actions, since this is in that realm I'd prefer to adopt their approach.

> Hiding execution under an import could make debugging more difficult later.

I'm not sure I understand the concern here - what is more difficult? If anything, it requires one fewer layer of indirection